Cabal/Distribution/Simple/GHC.hs

```diff
@@ -873,6 +874,43 @@ flibTargetName lbi flib =
     platformOS :: Platform -> OS
     platformOS (Platform _arch os) = os
 
+    -- If a foreign lib foo has elf-version 3.2.1, it should be built
+    -- as libfoo.so.3.2.1
+    versionedExt :: String
+    versionedExt = case foreignLibELFVersion flib of
+                     Nothing -> "so"
+                     Just v  -> "so" <.> display v
+
+-- | Name for the library when building.
+--
+-- If the `elf-version` field of a foreign library target is set, we
+-- need to incorporate that version into the SONAME field.
+--
+-- If a foreign library foo has elf-version 3.2.1, it should be built
+-- as libfoo.so.3.2.1.  We want it to get soname libfoo.so.3.
+-- However, GHC does not allow overriding soname by setting linker
+-- options, as it sets a soname of its own (namely the output
+-- filename), after the user-supplied linker options.  Hence, we have
+-- to compile the library with the soname as its filename.  We rename
+-- the compiled binary afterwards.
+--
+-- This method allows to adjust the name of the library at build time
+-- such that the correct soname can be set.
+flibBuildName :: LocalBuildInfo -> ForeignLib -> String
+flibBuildName lbi flib
+  | (platformOS (hostPlatform lbi), foreignLibType flib) ==
+    (Linux, ForeignLibNativeShared)
+  = case foreignLibELFVersion flib of
+      Nothing -> "lib" ++ nm <.> "so"
+      Just v  -> "lib" ++ nm <.> "so" <.> show (head (versionNumbers v))
```

Comment:
While I agree, it couldn't have resulted in errors, since a user can only set `v` to a non-null version.